pkg/kubectl/cmd/replace.go

```diff
@@ -296,3 +331,128 @@ func forceReplace(f cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []s
 	}
 	return nil
 }
+
+// If saveConfig is true, create or update the annotation.
+// If saveConfig is false and the local config file doesn't have the annotation, we save the annotation from the live object if there is one.
+// If saveConfig is false and the local config file has the annotation, we use the annotation in the config file.
+// info is the resource need to process. saveConfig indicate if we want to save the config in the annotation.
+func handleApplyAnnotation(info *resource.Info, saveConfig bool, lcl lastConfigurationLookup, encoder runtime.Encoder) error {
+	if saveConfig {
+		if err := kubectl.CreateOrUpdateAnnotation(true, info, encoder); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	anno, err := kubectl.GetOriginalConfiguration(info.Mapping, info.Object)
+	if err != nil {
+		return err
+	}
+	// The user-provided config doesn't have the annotation, try to get it from live object
+	if anno == nil {
+		annotationFromLiveObj, err := lcl.getAnnotationForResource(info)
+		if err != nil {
+			return err
+		}
+		if err = kubectl.SetOriginalConfiguration(info, annotationFromLiveObj); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// newGetResources is the factory method returning an object that implements lastConfigurationLookup interface
+func newGetResources(force bool, infos []*resource.Info) (lastConfigurationLookup, error) {
+	if force {
+		return buildMap(infos)
+	} else {
+		return lazyResources{}, nil
+	}
+}
+
+type lastConfigurationLookup interface {
+	getAnnotationForResource(info *resource.Info) ([]byte, error)
+}
+
+// lazyResources takes a list of resource that will be lazily read from the apiserver when they are needed.
+type lazyResources struct{}
+
+// getAnnotationForResource returns the annotation by fetching from the api server.
+func (lr lazyResources) getAnnotationForResource(info *resource.Info) ([]byte, error) {
+	// save the the local obj before the live obj overwrite it
+	localObj := info.Object
```

Comment:
I'm not a fan of the self-mutating and restoring happening here. This feels really complicated.